src/proxy/http/profiles.rs

```diff
@@ -475,33 +575,82 @@ pub mod router {
         }
     }
 
-    impl<G, T, R, B> Service<G, T, R, B>
+    impl<G, T, R, RMk, M, RBody, MBody>
+        Service<
+            G,
+            T,
+            R,
+            svc::shared::Shared<ConcreteRouter<T, M::Value, MBody>>,
+            RMk,
+            M,
+            RBody,
+            MBody,
+        >
     where
         G: Stream<Item = Routes, Error = Never>,
-        T: WithRoute + Clone,
+        T: WithRoute + WithAddr + Eq + Hash + Clone,
         T::Output: Clone + Eq + Hash,
-        R: rt::Make<T::Output> + Clone,
-        R::Value: svc::Service<http::Request<B>> + Clone,
+        R: svc::Layer<svc::shared::Shared<ConcreteRouter<T, M::Value, MBody>>, Service = RMk>
+            + Clone,
+        RMk: rt::Make<T::Output> + Clone,
+        RMk::Value: svc::Service<http::Request<RBody>> + Clone,
+        M: rt::Make<T> + Clone,
+        M::Value: svc::Service<http::Request<MBody>> + Clone,
     {
         fn update_routes(&mut self, routes: Routes) {
+            let capacity = routes.dst_overrides.len() + 1;
+
+            let mut make = IndexMap::with_capacity(capacity);
+            let mut drained = self
```

Comment:
Similar to my comment about finding a better name for `Router::drain()`, I think this could use a change in name as well. We don't use this value as an iterator -- just a lookup for targets that we expect to be there.